modules/@apostrophecms/log/index.js

```diff
@@ -0,0 +1,255 @@
+// is bundled into the second, object - based argument as a property of the name
+// given, and only the object argument is passed to Pino. If there is
+// no object-based argument an object is created.
+// Example:
+// ```js
+// {
+//   options: {
+//     messageAs: 'msg'
+//   }
+// }
+// ```
+// The resulting util log call will be:
+// ```js
+// self.apos.util.error({
+//   msg: '@apostrophecms/login: incorrect-username: User admin failed to log in',
+//   type: 'incorrect-username'
+//   module: '@apostrophecms/login'
+// });
+// ```
+//
+// ### `filter`
+//
+// By module name, or `*` we can specify any mix of severity levels and
+// specific event types, and entries are kept if *either* criterion is met.
+// Example:
+// ```js
+// {
+//   options: {
+//     filter: {
+//       // Log all errors and warnings from any module
+//       '*': {
+//         severity: [ 'warn', 'error' ]
+//       },
+//       // Log specific event types from the login module
+//       '@apostrophecms/login': {
+//         events: [ 'incorrect-username', 'incorrect-password' ]
+//       }
+//     }
+//   }
+// }
+// ```
+// In this example, all errors and warnings from any module, but
+// only the specific event types (no matter the severity) from the login
+// module, are logged.
+//
+// ## Environment Variables
+//
+// ### `APOS_FILTER_LOGS`
+//
+// If set, this environment variable overrides the `filter` option.
+// Example:
+// ```sh
+// # same as the `filter` example above
+// export APOS_FILTER_LOGS='*:severity:warn,error;@apostrophecms/login:events:incorrect-username,incorrect-password'
+// # log everything
+// export APOS_FILTER_LOGS='*'
+// ```
+//
+
+const _ = require('lodash');
+
+module.exports = {
+  options: {
+    alias: 'structuredLog'
+  },
+  init(self) {
+  },
+  methods(self) {
+    return {
+      // Logger for a module factory.
+      // The created logger object contains `debug`, `info`, `warn` and `error` methods,
+      // each suuporting the following signature:
+      // - (eventType)
+      // - (eventType, data)
+      // - (eventType, message)
+      // - (eventType, message, data)
+      // - (req, eventType[, message, data]) message and data are optional
+      getLoggerForModule(moduleSelf) {
+        return {
+          debug: (...args) => {
+            self.logEntry(moduleSelf, 'debug', ...args);
+          },
+          info: (...args) => {
+            self.logEntry(moduleSelf, 'info', ...args);
+          },
+          warn: (...args) => {
+            self.logEntry(moduleSelf, 'warn', ...args);
+          },
+          error: (...args) => {
+            self.logEntry(moduleSelf, 'error', ...args);
+          }
+        };
+      },
+      // Internal module, do not use it directly. See `@apostrophecms/module` for
+      // module level logging methods - logInfo, logError, etc.
+      logEntry(moduleSelf, severity, req, eventType, message, data) {
```

Comment:
when is moduleSelf ever missing?

Reply:
There was a case in the old implementation. Currently it's always provided. However it doesn't field good to force it at such low level.